File: Java/labs/lab20_State/task_2_1 – solution/src/MediaPlayer.java
public class MediaPlayer {

    private State state = new StatePause();
    private String icon = "play button";

    public void setState(State state) {
        this.state = state;
    }

    public State getState() {
        return state;
    }

    public String getIcon() {
        return icon;
    }

    public void setIcon(String icon) {
        this.icon = icon;
    }

    public void play() {
        state.play(this);
    }

    public void pause() {
        state.pause(this);
    }

}

File: Cpp/lab20_State/task_2_2/src/media_player.cpp
#include <iostream>

#include "media_player.hpp"


void MediaPlayer::setState(MediaPlayer::State state)
{
  this->state = state;
}

MediaPlayer::State MediaPlayer::getState() const
{
  return this->state;
}

void MediaPlayer::setIcon(std::string const& icon)
{
  this->icon = icon;
}

std::string const& MediaPlayer::getIcon() const
{
  return this->icon;
}

void MediaPlayer::play()
{
  switch (this->state)
  {
    case MediaPlayer::State::Paused:
    {
      this->setState(MediaPlayer::State::Playing);
      this->setIcon("pause button");

      std::cout << "Video playing, icon set to " << this->getIcon() << std::endl;
      break;
    }
    case MediaPlayer::State::Playing:
    {
      break;
    }
  }
}

void MediaPlayer::pause()
{
  switch (this->state)
  {
    case MediaPlayer::State::Paused:
    {
      break;
    }
    case MediaPlayer::State::Playing:
    {
      this->setState(MediaPlayer::State::Paused);
      this->setIcon("play button");

      std::cout << "Video paused, icon set to " << this->getIcon() << std::endl;
      break;
    }
  }
}
